File: frontend/types/tender-details.ts
import type {
    EngagementAction,
    EngagementOrigin,
    EngagementStatus,
} from '@/types/engagement';
import type { SourceSystem } from '@/types/tender';

export type DetailsSectionState = 'AVAILABLE' | 'EMPTY' | 'UNAVAILABLE';

export interface DetailsSection<T> {
    state: DetailsSectionState;
    data: T | null;
    reason_code: string | null;
}

export interface TenderDetailsProjectContext {
    project_id: string;
    external_project_id: string;
    name: string | null;
    source_system: SourceSystem;
    project_status: string | null;
    country: string | null;
    region: string | null;
    approval_date: string | null;
    closing_date: string | null;
    enrichment_state: string;
    last_enriched_at: string | null;
}

export interface TenderDetailsProjectLeadershipItem {
    role_id: string;
    role_type: 'PROJECT_LEADERSHIP';
    display_name: string;
    native_role: string;
    canonical_role: string;
    source_system: SourceSystem;
    source_url: string | null;
    is_current: boolean;
    first_observed_at: string;
    last_observed_at: string;
    ended_at: string | null;
}

export interface TenderDetailsProjectLeadership {
    items: TenderDetailsProjectLeadershipItem[];
    total_count: number;
    returned_count: number;
    truncated: boolean;
}

export interface TenderDetailsProcurementContacts {
    buyer_agency: string | null;
    contact_person: string | null;
    email: string | null;
    phone: string | null;
    address: string | null;
    submission_method: string | null;
    submission_deadline: string | null;
    question_deadline: string | null;
    procedure_type: string | null;
    participation_instructions: string | null;
    official_source_url: string | null;
    document_access_notes: string | null;
    source_type: 'TENDER_SOURCE';
}

export interface TenderDetailsRequirementItem {
    label: string;
    source_type: 'AI_EXTRACTED' | 'ANALYSIS_DERIVED';
    document_name: string | null;
    page: number | null;
    section: string | null;
}

export interface TenderDetailsRequirements {
    source_native_available: boolean;
    derivation: 'ANALYSIS_VERSION';
    items: TenderDetailsRequirementItem[];
    total_count: number;
    returned_count: number;
    truncated: boolean;
}

export type TenderDetailsDocumentAvailability =
    | 'AVAILABLE'
    | 'UNAVAILABLE'
    | 'METADATA_ONLY';

export interface TenderDetailsDocumentItem {
    document_id: string;
    display_name: string;
    document_type: string;
    metadata_classification: 'PUBLIC_SOURCE_METADATA';
    source_system: SourceSystem;
    availability: TenderDetailsDocumentAvailability;
    file_size: number | null;
    content_type: string | null;
    created_at: string;
}

export interface TenderDetailsDocuments {
    items: TenderDetailsDocumentItem[];
    visible_total_count: number;
    returned_count: number;
    omitted_unknown_count: number;
    truncated: boolean;
    download_authorization_separate: true;
}

export interface TenderDetailsCompliance {
    analysis_id: string;
    version_number: number;
    execution_state: string;
    compliance_completeness: string;
    decision_label: string | null;
    key_issue_count: number | null;
    coverage_signal: string | null;
    version_origin: string;
    override_applied: boolean;
    created_at: string;
    completed_at: string | null;
}

export interface TenderDetailsCompanyReadiness {
    profile_available: true;
    certifications_total: number;
    expired_certifications: number;
    licenses_total: number;
    active_licenses: number;
    credentials_total: number;
    expired_credentials: number;
    readiness_documents_total: number;
    readiness_documents_available: number;
    readiness_documents_missing: number;
    readiness_documents_expired: number;
    readiness_documents_unknown: number;
    financial_history_years: number;
}

export interface TenderDetailsPursuit {
    engagement_id: string;
    engagement_status: EngagementStatus;
    engagement_origin: EngagementOrigin;
    status_changed_at: string;
    allowed_actions: EngagementAction[];
}

export type ProposalStatus = 'DRAFT' | 'GENERATING' | 'COMPLETED' | 'SUBMITTED';

export interface TenderDetailsBidPreparation {
    proposal_id: string;
    proposal_status: ProposalStatus;
    created_at: string;
    detail_route_id: string;
}

export interface TenderDetailsResponse {
    tender_id: string;
    project_context: DetailsSection<TenderDetailsProjectContext>;
    project_leadership: DetailsSection<TenderDetailsProjectLeadership>;
    procurement_contacts: DetailsSection<TenderDetailsProcurementContacts>;
    requirements: DetailsSection<TenderDetailsRequirements>;
    documents: DetailsSection<TenderDetailsDocuments>;
    compliance: DetailsSection<TenderDetailsCompliance>;
    company_readiness: DetailsSection<TenderDetailsCompanyReadiness>;
    pursuit: DetailsSection<TenderDetailsPursuit>;
    bid_preparation: DetailsSection<TenderDetailsBidPreparation>;
}

File: backend/app/schemas/tender_details.py
"""Explicit, bounded schemas for the secondary Tender Details read model."""

from __future__ import annotations

from datetime import date, datetime
from enum import Enum
from typing import Literal
from uuid import UUID

from pydantic import BaseModel, Field

from app.models.base import ProposalStatus, TenderEngagementOrigin, TenderEngagementStatus


class DetailsSectionState(str, Enum):
    AVAILABLE = "AVAILABLE"
    EMPTY = "EMPTY"
    UNAVAILABLE = "UNAVAILABLE"


class ProjectContextSummary(BaseModel):
    project_id: UUID
    external_project_id: str
    name: str | None = None
    source_system: str
    project_status: str | None = None
    country: str | None = None
    region: str | None = None
    approval_date: date | None = None
    closing_date: date | None = None
    enrichment_state: str
    last_enriched_at: datetime | None = None


class ProjectLeadershipItem(BaseModel):
    role_id: UUID
    role_type: Literal["PROJECT_LEADERSHIP"] = "PROJECT_LEADERSHIP"
    display_name: str
    native_role: str
    canonical_role: str
    source_system: str
    source_url: str | None = None
    is_current: bool
    first_observed_at: datetime
    last_observed_at: datetime
    ended_at: datetime | None = None


class ProjectLeadershipSummary(BaseModel):
    items: list[ProjectLeadershipItem] = Field(default_factory=list)
    total_count: int
    returned_count: int
    truncated: bool


class ProcurementContactsSummary(BaseModel):
    buyer_agency: str | None = None
    contact_person: str | None = None
    email: str | None = None
    phone: str | None = None
    address: str | None = None
    submission_method: str | None = None
    submission_deadline: datetime | None = None
    question_deadline: datetime | None = None
    procedure_type: str | None = None
    participation_instructions: str | None = None
    official_source_url: str | None = None
    document_access_notes: str | None = None
    source_type: Literal["TENDER_SOURCE"] = "TENDER_SOURCE"


class RequirementSummaryItem(BaseModel):
    label: str
    source_type: Literal["AI_EXTRACTED", "ANALYSIS_DERIVED"]
    document_name: str | None = None
    page: int | None = None
    section: str | None = None


class RequirementsSummary(BaseModel):
    source_native_available: bool = False
    derivation: Literal["ANALYSIS_VERSION"] = "ANALYSIS_VERSION"
    items: list[RequirementSummaryItem] = Field(default_factory=list)
    total_count: int
    returned_count: int
    truncated: bool


class TenderDocumentSummaryItem(BaseModel):
    document_id: UUID
    display_name: str
    document_type: str
    metadata_classification: Literal["PUBLIC_SOURCE_METADATA"]
    source_system: str
    availability: Literal["AVAILABLE", "UNAVAILABLE", "METADATA_ONLY"]
    file_size: int | None = None
    content_type: str | None = None
    created_at: datetime


class TenderDocumentsSummary(BaseModel):
    items: list[TenderDocumentSummaryItem] = Field(default_factory=list)
    visible_total_count: int
    returned_count: int
    omitted_unknown_count: int
    truncated: bool
    download_authorization_separate: Literal[True] = True


class ComplianceSummary(BaseModel):
    analysis_id: UUID
    version_number: int
    analysis_language: str | None = None
    execution_state: str
    compliance_completeness: str
    decision_label: str | None = None
    key_issue_count: int | None = None
    coverage_signal: str | None = None
    version_origin: str
    override_applied: bool
    created_at: datetime
    completed_at: datetime | None = None


class CompanyReadinessSummary(BaseModel):
    profile_available: Literal[True] = True
    certifications_total: int
    expired_certifications: int
    licenses_total: int
    active_licenses: int
    credentials_total: int
    expired_credentials: int
    readiness_documents_total: int
    readiness_documents_available: int
    readiness_documents_missing: int
    readiness_documents_expired: int
    readiness_documents_unknown: int
    financial_history_years: int


class PursuitSummary(BaseModel):
    engagement_id: UUID
    engagement_status: TenderEngagementStatus
    engagement_origin: TenderEngagementOrigin
    status_changed_at: datetime
    allowed_actions: list[str] = Field(default_factory=list)


class BidPreparationSummary(BaseModel):
    proposal_id: UUID
    proposal_status: ProposalStatus
    created_at: datetime
    detail_route_id: UUID


class ProjectContextSection(BaseModel):
    state: DetailsSectionState
    data: ProjectContextSummary | None = None
    reason_code: str | None = None


class ProjectLeadershipSection(BaseModel):
    state: DetailsSectionState
    data: ProjectLeadershipSummary | None = None
    reason_code: str | None = None


class ProcurementContactsSection(BaseModel):
    state: DetailsSectionState
    data: ProcurementContactsSummary | None = None
    reason_code: str | None = None


class RequirementsSection(BaseModel):
    state: DetailsSectionState
    data: RequirementsSummary | None = None
    reason_code: str | None = None


class TenderDocumentsSection(BaseModel):
    state: DetailsSectionState
    data: TenderDocumentsSummary | None = None
    reason_code: str | None = None


class ComplianceSection(BaseModel):
    state: DetailsSectionState
    data: ComplianceSummary | None = None
    reason_code: str | None = None


class CompanyReadinessSection(BaseModel):
    state: DetailsSectionState
    data: CompanyReadinessSummary | None = None
    reason_code: str | None = None


class PursuitSection(BaseModel):
    state: DetailsSectionState
    data: PursuitSummary | None = None
    reason_code: str | None = None


class BidPreparationSection(BaseModel):
    state: DetailsSectionState
    data: BidPreparationSummary | None = None
    reason_code: str | None = None


class TenderDetailsResponse(BaseModel):
    tender_id: UUID
    project_context: ProjectContextSection
    project_leadership: ProjectLeadershipSection
    procurement_contacts: ProcurementContactsSection
    requirements: RequirementsSection
    documents: TenderDocumentsSection
    compliance: ComplianceSection
    company_readiness: CompanyReadinessSection
    pursuit: PursuitSection
    bid_preparation: BidPreparationSection
